File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque 
def solution(begin, target, words):
    answer = 0
    queue = deque([(begin,0)])
    visited = [] # 한번 방문시 다시는 방문 못하게 
    while queue:
        begin_word,count = queue.popleft()
        
        # 종료 시점 
        if begin_word == target:
            answer = count
            break 
        
        for word in words:
            if word in visited:
                continue
            same_count = 0 
            for bw, w in zip(list(begin_word),list(word)):
                if bw == w :
                    same_count+=1 
            
            # 큐에 추가 
            if same_count == len(word)-1 :
                queue.append([word,count+1])
                visited.append(word)
                    
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    private boolean canConvert(String a, String b) {
        int diff = 0;
        for (int i = 0; i < a.length(); i++) {
            if (a.charAt(i) != b.charAt(i) && ++diff > 1) return false;
        }
        return diff == 1;
    }

    public int solution(String begin, String target, String[] words) {
        boolean exist = false;
        for (String w : words) if (w.equals(target)) { exist = true; break; }
        if (!exist) return 0;

        boolean[] visited = new boolean[words.length];
        Queue<int[]> q = new LinkedList<>(); // [index, depth], begin은 -1
        q.offer(new int[]{-1, 0});

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            String curWord = cur[0] == -1 ? begin : words[cur[0]];
            int depth = cur[1];

            if (curWord.equals(target)) return depth;

            for (int i = 0; i < words.length; i++) {
                if (visited[i]) continue;
                if (canConvert(curWord, words[i])) {
                    visited[i] = true;
                    q.offer(new int[]{i, depth + 1});
                }
            }
        }
        return 0;
    }
}